src/js/Menus/Menu.js

```diff
@@ -156,133 +297,136 @@ export default class Menu extends PureComponent {
+    const { cascading, id, fixedTo, cascadingAnchor } = this.props;
+    const listLevel = this.context.listLevel || 0;
+    const cascadingMenu = typeof cascading !== 'undefined' ? cascading : this.context.cascadingMenu;
+    const cascadingZDepth = getField(this.context, this.props, 'cascadingZDepth');
+    const cascadingFixedTo = typeof fixedTo !== 'undefined' ? fixedTo : this.context.cascadingFixedTo;
+
     return {
-      menuCascading,
-      menuPosition,
-      listLevel: 0,
+      listLevel,
+      cascadingId: `${id}-level-${listLevel + 1}`,
+      cascadingMenu,
+      cascadingAnchor,
+      cascadingZDepth,
+      cascadingFixedTo,
     };
   }
 
   componentDidMount() {
-    const { isOpen } = this.props;
-    if (isOpen) {
-      window.addEventListener('click', this._handleOutsideClick);
-    }
-  }
-
-  componentDidUpdate(prevProps) {
-    const { isOpen } = this.props;
-    if (isOpen === prevProps.isOpen) {
-      return;
-    }
-
-    window[`${isOpen ? 'add' : 'remove'}EventListener`]('click', this._handleOutsideClick);
+    this._container = findDOMNode(this);
   }
 
   componentWillUnmount() {
-    const { isOpen } = this.props;
-    if (isOpen) {
-      window.removeEventListener('click', this._handleOutsideClick);
-    }
-
     if (this._timeout) {
       clearTimeout(this._timeout);
     }
   }
 
-  _setList(list) {
-    if (list !== null) {
-      this._list = findDOMNode(list);
-    }
-
-    try {
-      const children = Children.only(this.props.children);
-      if (typeof children.ref === 'function') {
-        children.ref(list);
-      }
-    } catch (e) {
-      // do nothing
-    }
-  }
-
-  _setContainer(container) {
-    if (container !== null) {
-      this._container = findDOMNode(container);
+  _handleClose = (e) => {
+    const { close, onClose } = this.props;
+    if (close || onClose) {
+      (close || onClose)(e);
     }
-  }
-
-  _handleOutsideClick(e) {
-    const isInContextMenu = this.props.position === Positions.CONTEXT
-      && !this._list.contains(e.target);
+  };
 
-    if (isInContextMenu || !this._container.contains(e.target)) {
-      const { onClose, close } = this.props;
-      if (close) {
-        close(e);
-      } else if (onClose) {
-        onClose(e);
-      }
+  _handleClick = (e) => {
+    if (this.props.onClick) {
+      this.props.onClick(e);
     }
-  }
-
-  _handleListClick(e) {
-    const { onClose, close, autocloseAfterInk } = this.props;
 
     let node = e.target;
-    while (this._container.contains(node)) {
-      if (!node.classList.contains('md-list-item--nested-container') && node.classList.contains('md-list-item')) {
+    while (this._container && this._container.contains(node)) {
+      if (node.classList.contains('md-list-control')) {
+        return;
+      } else if (
+        !node.classList.contains('md-list-item--nested-container') &&
+        node.classList.contains('md-list-item')
+      ) {
         this._timeout = setTimeout(() => {
```

Comment:
Add comment about how this allows any click events on those list item to be triggered before closing.